### PR TITLE
 Monster: Cleanup/Refactor http controllers.

### PR DESCRIPTION
**Pull request by mnemonicly**
This cleans up docblocks, pulls most non-crudy actions into their own
controllers, and does general cleanup/logic refactoring.  There /should/
be no functional changes, but we all know how should works..

I'm happy to split this up into multiple smaller PR's if that is going to be cleaner to review.

### CODE REVIEW

**Comment by snipe** [comment]
Merging that license seat PR gave us a merge conflict - can you resolve or should I?

**Comment by dmeltzer** [comment]
Resolved.  Also cleaned that method up a bit more... I think it makes sense now?

File: app/Http/Controllers/Consumables/ConsumableCheckoutController.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        $consumable->assigned_to = e(Input::get('assigned_to'));
+
+        $consumable->users()->attach($consumable->id, [
+            'consumable_id' => $consumable->id,
+            'user_id' => $admin_user->id,
+            'assigned_to' => e(Input::get('assigned_to'))
+        ]);
+
+        $logaction = $consumable->logCheckout(e(Input::get('note')), $user);
+        $data['log_id'] = $logaction->id;
```
Comment (uberbrady) [inline app/Http/Controllers/Consumables/ConsumableCheckoutController.php]: Where's this data end up? It also doesn't look used.

**Review by uberbrady — COMMENTED** [review]
I'm feeling kinda ambivalent, maybe because I have something different in my head for how to clean some of this stuff up.

But I *do* very much like the fact that each checkin/checkout controller has a `create` and then a `store` method - that makes it certainly start to _feel_ a little more like they deserve their own controllers for these things.

My initial instinct is that checkin and checkout should just be one grand unified controller but then you do certainly lose the kinda REST-ish `create` and `store` methods that are cleanly one-per-file.

I originally came to Laravel from a Rails background, so I'm used to the 'fat model, skinny controller' paradigm. So my thought for how I wanted to see this cleaned up was this: I wanted more of the checkin/checkout logic that applies to the specific things that can be checked out or in to be embedded in the various things themselves - and then have the controller methods be much thinner, containing less asset-specific or accessory-specific or component-specific logic in it. If we could keep the 'smarts' embedded well enough in the model.

This could - eventually - get us to a "Checkoutable" trait that we can apply to all things that can be checked out, which would make it easier for us to really start to get rid of more duplicated code.

It *might* be possible to create a generic checkin/checkout controller that just calls `$thing->checkout($recipient)` or `$thing->checkin($optional_checkin_to_where)` - and then be able to just re-apply that generic controller across all the checkin/checkout types.

*That* would definitely be something I could get behind. We'd probably have to use some kind of polymorphism to figure out which view to show, but, still - that seems like it'd get rid of a lot of code, and could really add some value here.

That being said, when I first saw this I didn't like it - I do like my things to be in one file so I don't have, like, 15 files open when I'm trying to troubleshoot something, for example. But after having looked closer I don't hate it anywhere near as much. I could really see it going either way.

**Comment by tilldeeke** [comment]
I personally found that thinking in “resources” often led to a cleaner code flow or layout. So I’m all for extracting methods into their separate controllers if they operate on the same resource.

> My initial instinct is that checkin and checkout should just be one grand unified controller but then you do certainly lose the kinda REST-ish create and store methods that are cleanly one-per-file.

You could alternatively view the checkin/checkout flow as _storing_ an assignment resource and _destroying_ that resource. That way you could use one controller for checkin/checkout and would get semantic meaning behind the REST-ish method names.

> That being said, when I first saw this I didn't like it - I do like my things to be in one file so I don't have, like, 15 files open when I'm trying to troubleshoot something, for example.

I'd argue that extracting these methods helps in troubleshooting, since it trims down the number of lines of code you'd have to scan to find problems.

File: app/Http/Controllers/GroupsController.php [hunk 2]
```diff
@@ -152,19 +146,19 @@ public function destroy($id = null)
...
     {
-        $group = Group::find($id);
+        $group = Group::find($groupId);
 
         if ($group) {
             return view('groups/view', compact('group'));
         }
 
-        return redirect()->route('groups.index')->with('error', trans('admin/groups/message.group_not_found', compact('id')));
+        return redirect()->route('groups.index')->with('error', trans('admin/groups/message.group_not_found', compact('groupId')));
```
[the remark this answers is not in the document]
Comment (dmeltzer) [reply app/Http/Controllers/GroupsController.php]: Good Catch.. There's actually a number of these.  If it's any better, the old way was wrong too :) (id was never defined.)  Codacy caught it, I just did a bad job of fixing it.

File: app/Http/Controllers/Accessories/AccessoryCheckoutController.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        $data['log_id'] = $logaction->id;
+        $data['eula'] = $accessory->getEula();
+        $data['first_name'] = $user->first_name;
+        $data['item_name'] = $accessory->name;
+        $data['checkout_date'] = $logaction->created_at;
+        $data['item_tag'] = '';
+        $data['expected_checkin'] = '';
+        $data['note'] = $logaction->note;
+        $data['require_acceptance'] = $accessory->requireAcceptance();
```
[the remark this answers is not in the document]
Comment (dmeltzer) [reply app/Http/Controllers/Accessories/AccessoryCheckoutController.php]: Ah yes, I think this is left over from when the mail notification of a checkout was done in the controller.  This was extracted to an event a while back but I think the array was never removed.

**Comment by dmeltzer** [comment]
> My initial instinct is that checkin and checkout should just be one grand unified controller but then you do certainly lose the kinda REST-ish create and store methods that are cleanly one-per-file.

I went back and forth about how to handle checkin/checkout, and whether they should be one controller or two.  I ended up sticking with two controllers because it felt cleaner/more predictable to have a create and store method than to have a create, store, getCheckin()? and destroy method.  I'm convincable to go in either direction, however.  

> I originally came to Laravel from a Rails background, so I'm used to the 'fat model, skinny controller' paradigm. So my thought for how I wanted to see this cleaned up was this: I wanted more of the checkin/checkout logic that applies to the specific things that can be checked out or in to be embedded in the various things themselves - and then have the controller methods be much thinner, containing less asset-specific or accessory-specific or component-specific logic in it. If we could keep the 'smarts' embedded well enough in the model.

This is also a goal of mine, especially with the current regular semi-duplication between Controller(store|update) and api/Controller(store|update).  Would love to look at extracting this either to the model (although those are already fairly big and bloaty) or where it works, a formrequest class that encapsulates the common logic.  We have Asset::checkOut() currently, but thats the only place the checkout logic does not occur in the controller.  

> That being said, when I first saw this I didn't like it - I do like my things to be in one file so I don't have, like, 15 files open when I'm trying to troubleshoot something, for example. But after having looked closer I don't hate it anywhere near as much. I could really see it going either way.

I've gone back and forth on this as well.  While I'm not a huge fan of having many files open, I'm less of a fan of having to scroll through 900 lines of a file to find the appropriate function, (although in reality, modern editors help with both of these problems).  I think that organization limitations is one of my current biggest gripes with php and javascript, and I'm still trying to figure out the best way to keep things organized.

**Comment by dmeltzer** [comment]
Fixed most of the concerns @uberbrady raised.

**Pull request merged**